### PR TITLE
Fixed a bug where msal-node doesn't work behind a proxy

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**

### CODE REVIEW

File: lib/msal-node/src/network/HttpClient.ts [hunk 1]
```diff
@@ -23,21 +25,39 @@ export class HttpClient implements INetworkModule {
...
+        proxy?: string
     ): Promise<NetworkResponse<T>> {
-        const request: AxiosRequestConfig = {
-            method: HttpMethod.GET,
-            url: url,
-            headers: options && options.headers,
-            validateStatus: () => true
-        };
+        let response;
+        let responseObject: any = {};
```
Comment (jasonnutter) [inline lib/msal-node/src/network/HttpClient.ts:33]: Prefer to not use `any`, please define a type.

File: lib/msal-node/src/network/HttpClient.ts [hunk 2]
```diff
@@ -23,21 +25,39 @@ export class HttpClient implements INetworkModule {
      */
     async sendGetRequestAsync<T>(
         url: string,
-        options?: NetworkRequestOptions
+        options?: NetworkRequestOptions,
+        cancellationToken?: number,
+        proxy?: string
```
Comment (jasonnutter) [inline lib/msal-node/src/network/HttpClient.ts:30]: Can `proxy` and `cancellationToken` be added to `NetworkRequestOptions` instead? I would like us to get out of the habit of many optional parameters like this.

File: lib/msal-node/src/network/HttpClient.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ import {
 } from "@azure/msal-common";
 import { HttpMethod } from "../utils/Constants";
 import axios, { AxiosRequestConfig } from "axios";
+import { HttpsProxyAgent } from "https-proxy-agent";
+import fetch from "node-fetch";
```
Comment (jasonnutter) [inline lib/msal-node/src/network/HttpClient.ts:14]: So Axios doesn't support working with proxies? I find that surpising.

File: lib/msal-node/src/network/HttpClient.ts [hunk 4]
```diff
@@ -23,21 +25,39 @@ export class HttpClient implements INetworkModule {
...
+        let responseObject: any = {};
+
+        if (proxy) {
+            const fetchOptions =  {
+                headers: options?.headers,
+                body: options?.body,
+                agent: new HttpsProxyAgent(proxy),
+            };
+    
+            response = await fetch(url, fetchOptions);
```
Comment (jasonnutter) [inline lib/msal-node/src/network/HttpClient.ts:42]: The other option is to document how applications can provide their own http client, which give them complete control over the proxied requests. Did we consider that?

File: lib/msal-node/src/network/HttpClient.ts (hunk 1 above)
Comment (jasonnutter) [reply lib/msal-node/src/network/HttpClient.ts:33]: And `const` is fine, since you are not reassigning the object.

**Review by jasonnutter — CHANGES_REQUESTED** [review]
A few comments. Also, would like to see tests for this feature. Thanks!

File: lib/msal-node/src/network/HttpClient.ts (hunk 3 above)
Comment (sameerag) [reply lib/msal-node/src/network/HttpClient.ts:14]: Yeah. Its took a few escalations before we figured axios is the issue. :)

File: lib/msal-node/src/network/HttpClient.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-node/src/network/HttpClient.ts:42]: Yes. I am about to suggest that. The alternative I have in mind is provide a sample with a network client that supports proxies and leave it at that with instructions. We are open to suggestions.

Comment (Robbie-Microsoft) [reply lib/msal-node/src/network/HttpClient.ts]: Per our discussion, we are switching from axios to node-fetch.

**Comment by tnorling** [comment]
Addressed in #4447

**Pull request closed without merging**